doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,331 @@
+      - Also, there could be a sentry warning with the above soft assert, so that we know if a customer is probably running into a performance problem due to falling back to the old window function implementation.
+- Performance testing: Importantly, we need to test that we efficiently support situations when small input changes lead to small output changes.
+    - Writing automated performance tests is tricky though. Currently, we don’t have any automated performance tests.
+    - At least manual testing should definitely be performed before merging the PRs, since the whole point of this work is performance.
+    - We could do it roughly as follows:
+        - We put in lots of input data with one timestamp, as an “initial snapshot”. Processing this should be at least several seconds.
+        - We change a small portion of the input data.
+            - But the total size of the affected partitions should cover most of the input data. This is important, since if window partitions are very small, then the current window function support works fine.
+            - This should complete orders of magnitude faster than the initial snapshot.
+    - More specifically, maybe we could have a Testdrive test that performs the following steps:
+        - Copy some TPC-H data from our TPC-H source into tables.
+        - Create a materialized view with some window functions on the tables.
+        - Do some inserts/updates/deletes on the tables.
+        - Check that updating of the materialized view happens quickly.
+            - Should be possible to set up the input data and the queries in such a way that
+                - updating takes orders of magnitude faster than the initial snapshot.
+                - But not with the current window function support.
+                - The difference from the initial snapshot should be big enough so that the test won’t be flaky.
+    - (I already tested a [simple prototype for LAG outside Materialize](https://github.com/ggevay/window-funcs).)
+- We should measure the memory needs of Prefix Sum, so that we can advise users when sizing replicas.
+
+## Lifecycle
+
+There are many window functions, and many frame options. We will gradually add the new, efficient implementations for an increasing set of window function + frame setting combinations across several PRs. We will make sure the new implementation is correct and performs well before merging each of the PRs.
+
+# Open questions
+
+- How to have automated performance tests? Can we check in Testdrive that some materialized view (that has window functions) is being updated fast enough?
```

Comment:
Since we don't have automated performance tests, I would not block on it. However, we need manual performance tests to validate this design.

Reply:
Ok, I clarified that it's ok to only have manual performance tests for the first version.